x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java

```diff
@@ -191,14 +171,33 @@ private SearchResponse findOrBuildResponse(AsyncSearchTask task) {
 
     /**
      * Creates an {@link AsyncSearchResponse} based on the current state of the mutable response.
-     * This method also restores the response headers in the current thread context if the final response is available.
+     * The final reduce of the aggregations is executed if needed (partial response).
+     * This method is synchronized to ensure that we don't perform final reduces concurrently.
+     * This method also restores the response headers in the current thread context when requested, if the final response is available.
      */
-    synchronized AsyncSearchResponse toAsyncSearchResponseWithHeaders(AsyncSearchTask task, long expirationTime) {
-        AsyncSearchResponse resp = toAsyncSearchResponse(task, expirationTime);
-        if (responseHeaders != null) {
+    synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task,
+                                                           long expirationTime,
+                                                           boolean restoreResponseHeaders) {
+        if (restoreResponseHeaders && responseHeaders != null) {
             restoreResponseHeadersContext(threadContext, responseHeaders);
         }
-        return resp;
+        return new AsyncSearchResponse(task.getExecutionId().getEncoded(), findOrBuildResponse(task),
+            failure, isPartial, frozen == false, task.getStartTime(), expirationTime);
+    }
+
+    synchronized AsyncSearchResponse buildErrorResponse(AsyncSearchTask task,
+                                                        long expirationTime,
+                                                        ElasticsearchException exception) {
+        ElasticsearchException error;
+        if (this.failure == null) {
+            error = exception;
+        } else {
+            error = this.failure;
+            error.addSuppressed(exception);
+        }
+        //TODO add some search response here rather than null
```

Comment:
++, we should return the current stats (number of shards, shard failures, ...) and just omit the partial aggs.